### PR TITLE
Ensure the correct MPIEXEC is used

### LINKED ISSUE #359: Defect: CMake can't find mpiexec if not in PATH
| Avg response time              |
| ------------------------------ |
| ![Issue Stats][issue stat img] |

 - [X] I have surrounded any code, or log output in codeblock fences
       (` ``` `) if it occupies more than one line, or in single
       backticks if it is short `short code`

# Defect/bug report #

With the current master branch (OpenCoarrays 1.8.5) and MPICH 3.1.4, although though [line 27] in the build_opencoarrays.sh script sets the CC and FC environment variables and passes the MPI_C_COMPILER and MPI_Fortran_COMPILER arguments, CMake reports that it can't find mpiexec (despite finding mpif90)) unless mpiexec is in the PATH:
```bash
CC="${CC}" FC="${FC}" $CMAKE "${opencoarrays_src_dir}" \
  -DCMAKE_INSTALL_PREFIX="${install_path}" \
  -DMPI_C_COMPILER="${MPICC}" \
  -DMPI_Fortran_COMPILER="${MPIFC}"
```
The above behavior happens with our default CMake 3.4.0.  I'm investigating whether updating to a newer CMake version would resolve this or whether we simply need to set add the MPICH bin directory to the PATH before the above statement executes (which seems like overkill).   Depending on the outcome of my investigations, I'll either edit the build_opencoarrays.sh script to adjust the PATH before the above statement executes or I'll bump the minimum CMake version number.  I'll also look into bumping the MPICH version number as requested in issue #355.

The title of the issue should start with `Defect:` followed by a
succinct title.

 - [ ] I have searched the [issues page] and [mailing list] and did
       not find any issue matching the one I would like to report
 - [X] I have included a succinct description of the problem,
       including the steps necessary to reproduce it
 - [X] I have included which version of OpenCoarrays I am working
       with, or the output of `git describe` if working with a cloned
       repository rather than an [official release]
 - [X] I have included the MPI library name and version I am using
       with OpenCoarrays
   - [ ] name: [MPICH] version: 3.1.4
 - [X] I have included the Fortran compiler name and version that I am
       using with OpenCoarrays: 7.0.1
 - [ ] I have included the output of `uname -a`:  
`Linux rouson-VirtualBox 4.8.0-22-generic #24-Ubuntu SMP Sat Oct 8 09:15:00 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`
 - [X] I have included the Operating system name and version,
       including linux distro: Lunbuntu 16.10
 - [X] I have included the machine name if I am allowed to do so and
       wish to seek help debugging on that particular system (Titan,
       Pleiades, Excalibur, Lightening, Mira, etc.): MacBook Pro

## Installation/build problem ##

**Please select one from the following top level items** (please be sure to fill out all sub-items)**:**
 - [X] Build/install was attempted with [install.sh]
    - [ ] I have attached the `install-opencoarrays.log` file
          after re-running the script with the debug flags `-d` and
          `-v` turned on (after renaming the log file to have a `.txt`
          extension)
 - [ ] Build/install was attempted with system package manager like
       [homebrew] (which one?)
 - [ ] Build/install was attempted with manual/advanced
       [CMake install]
   - [X] Name and version of the C compiler being used: gcc
     - [X] Name and version of the Fortran compiler being used: gfortran
     - [X] Attached/included the output of the configure step with
           CMake. (Usually something like `cd build && cmake ..`):  see above
     - [ ] Attached/included the output of `make VERBOSE=1`
     - [ ] Attached/included the output of `ctest --verbose`
 - [X] Other build/install was attempted (please describe): macOS build under way

## Compile time or run time error

**Please choose one of the first two items**
 - [X] I am experiencing a compile time error (all gfortran ICEs
       should be reported to the [GFortran bug tracker]... You can post
       the ICE here too for tracking purposes, with a link to the
       GFortran problem report)
 - [ ] I am experiencing a runtime issue

**:sparkles: Optional, but you will receive infinite karma for including this: :sparkles:**
 - [X] I have included a minimally complete verifiable example (MCVE)
       that exhibits the problem and suitable for adding to regression
       tests: the example is the new [patched-trunk-install.sh] script.

[links]:#
[GFortran bug tracker]: https://gcc.gnu.org/bugzilla/
[Intel MPI]: https://software.intel.com/en-us/intel-mpi-library
[MPT]: http://www.sgi.com/products/software/sps.html
[MVAPICH]: http://mvapich.cse.ohio-state.edu
[MPICH]: https://www.mpich.org
[Open-MPI]: https://www.open-mpi.org
[CONTRIBUTING.md]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md
[Helping Out]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md#helping-out
[official release]: https://github.com/sourceryinstitute/OpenCoarrays/releases
[CMake install]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/INSTALL.md#cmake-scripts
[homebrew]: http://brew.sh
[issues page]: https://github.com/sourceryinstitute/OpenCoarrays/issues
[mailing list]: https://groups.google.com/forum/#!forum/opencoarrays
[install.sh]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/install.sh
[issue stat img]:https://img.shields.io/issuestats/i/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square
[lin 27]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/prerequisites/install-functions/build_opencoarrays.sh#L27
[patched-trunk-install.sh]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/developer-scripts/patched-trunk-install.sh

### PR DESCRIPTION
**Pull request by zbeekman**
Fixes #359

| Avg response time                 | coverage on master         |
| --------------------------------- | ---------------------------|
| ![Issue Stats][PR response img]   | ![Codecov branch][coverage]|

## Summary of changes ##

Hunt for `mpiexec` or `mpirun` in the same directory that `MPICC` and `MPIFC` are in when building with `install.sh` and then pass `-DMPIEXEC=...` to CMake so that the proper runtime environment is used...

## Rationale for changes ##

In some cases mpiexec was getting pulled in from the user's `PATH` and differed from the MPI implementation selected by the build script. This caused incompatibilities since `mpiexec`/`mpirun` from a different MPI implementation than the one used to compile was  being used.

## Additional info and certifications ##

This pull request (PR) is a:
 - [x] Bug fix
 - [ ] Feature addition
 - [ ] Other, Please describe:

I certify that:

 - [x] I have reviewed the [contributing guidelines]
 - [x] If this PR is a work in progress I have added `WIP:` to the
       beginning of the PR title
 - [x] If this PR is problematic for any reason, I have added
       `DO NOT MERGE:` to the beginning of the title
 - [x] The branch name and title of this PR contains the text
       `issue-<#>` where `<#>` is replaced by the issue that this PR
       is addressing
 - [x] I have deleted trailing white space on any lines that this PR
       touches
 - [x] I have used spaces for indentation on any lines that this PR
       touches
 - [x] I have included some comments to explain non-obvious code
       changes
 - [x] I have run the tests localy (`ctest`) and all tests pass
 - [x] Each commit is a logically atomic, self-consistent, cohesive
       set of changes
 - [x] The [commit message] should follow [these guidelines]:
     - [x] First line is directive phrase, starting with a capitalized
           imperative verb, and is no longer than 50 characters
           summarizing your commit
     - [x] Next line, if necessary is blank
     - [x] Following lines are all wrapped at 72 characters and can
           include additional paragraphs, bulleted lists, etc.
     - [x] Use [Github keywords] where appropriate, to indicate the
           commit resolves an open issue.
 - [x] I have signed  [Contributor License Agreement (CLA)] by
       clicking the "details" link to the right of the `licence/cla`
       check and following the directions on the CLA assistant webpage
 - [x] I have ensured that the test coverage hasn't gone down and added new [unit tests] to cover an new code added to the library

## For contributors and SI team members with code review priviledges ##

 - [x] I certify that I will wait 24 hours before self-approving via
       [pullapprove comment] or [Github code review] so that someone
       else has the chance to review my proposed changes

[links]:#
[contributing guidelines]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md
[commit message]: https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message
[these guidelines]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[Contributor License Agreement (CLA)]: https://cla-assistant.io/sourceryinstitute/OpenCoarrays
[pullapprove comment]: https://pullapprove.com/sourceryinstitute/OpenCoarrays/settings/
[Github code review]: https://help.github.com/articles/about-pull-request-reviews/
[Github keywords]: https://help.github.com/articles/closing-issues-via-commit-messages/
[unit tests]: https://github.com/sourceryinstitute/OpenCoarrays/tree/master/src/tests/unit
[PR response img]: https://img.shields.io/issuestats/p/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square
[coverage]: https://img.shields.io/codecov/c/github/sourceryinstitute/OpenCoarrays/master.svg?style=flat-square

### CODE REVIEW

**Comment by zbeekman** [comment]
This PR is ready to ship! I have been able to reproduce the bug both on MacOS and Lubuntu virtual machine.

 b855f89 ensures that a warning is produced if there is a mismatch between the MPI environment, or an error if `MPIEXEC` is missing completely, which should help prevent or catch #359 in the future.

This PR closes #360

File: CMakeLists.txt [hunk 1]
```diff
@@ -155,7 +155,7 @@ endif()
 
 find_package( MPI )
 
-if ( (NOT MPI_C_FOUND) OR (NOT MPI_Fortran_FOUND) )
+if ( (NOT MPI_C_FOUND) OR (NOT MPI_Fortran_FOUND) OR (NOT MPIEXEC))
```
Comment (zbeekman) [inline CMakeLists.txt]: This will fallback to the `install.sh` default installation location for MPICH.

File: CMakeLists.txt [hunk 2]
```diff
@@ -164,6 +164,46 @@ if ( (NOT MPI_C_FOUND) OR (NOT MPI_Fortran_FOUND) )
   find_package( MPI REQUIRED )
 endif()
 
+# Test for consistent MPI environment
+if (NOT MPIEXEC)
+  message ( ERROR "CMake failed to find `mpiexec` or similar. If building with `./install.sh` please
+report this bug to the OpenCoarrays developers at
```
Comment (zbeekman) [inline CMakeLists.txt]: This should throw an error if MPIEXEC is not found, and catch #359 or similar in the future

File: CMakeLists.txt [hunk 3]
```diff
@@ -164,6 +164,46 @@ if ( (NOT MPI_C_FOUND) OR (NOT MPI_Fortran_FOUND) )
...
+
+get_filename_component(MPIFC_RELATIVE_LOC "${MPI_Fortran_COMPILER}"
+  PROGRAM)
+get_filename_component(MPIFC_ABS_LOC "${MPIFC_RELATIVE_LOC}"
+  REALPATH)
+get_filename_component(MPIFC_DIR "${MPIFC_ABS_LOC}"
+  DIRECTORY)
+
+if ((MPIEXEC_DIR STREQUAL MPICC_DIR) AND (MPIEXEC_DIR STREQUAL MPIFC_DIR))
+  message ( STATUS "MPI runtime and compile time environments appear to be consistent")
```
Comment (zbeekman) [inline CMakeLists.txt]: If `MPIEXEC` is not from the same dir as `MPICC` then issue a warning.

File: prerequisites/install-functions/build_opencoarrays.sh [hunk 4]
```diff
@@ -22,9 +22,16 @@ build_opencoarrays()
   FC="${MPIFC_show[0]}"
   # Set CC to the MPI implementation's gcc command...
   CC="${MPICC_show[0]}"
+  # try to find mpiexec
+  MPIEXEC_CANDIDATES=($(find "${MPICC%/*}" -name 'mpiexec' -o -name 'mpirun'))
```
Comment (zbeekman) [inline prerequisites/install-functions/build_opencoarrays.sh]: Assumes MPICC set correctly, and searches for `mpiexec` or similar in same directory as compiler wrapper.

File: prerequisites/install-functions/build_opencoarrays.sh [hunk 5]
```diff
@@ -22,9 +22,16 @@ build_opencoarrays()
...
+  if ! ((${#MPIEXEC_CANDIDATES[@]} >= 1)); then
+    emergency "Could not find a suitable \`mpiexec\` in directory containing mpi wrappers (${MPICC%/*})"
+  else
+    MPIEXEC="${MPIEXEC_CANDIDATES[0]}"
```
Comment (zbeekman) [inline prerequisites/install-functions/build_opencoarrays.sh]: Pick out first `mpiexec`-like launcher we find

File: prerequisites/install-functions/build_opencoarrays.sh [hunk 6]
```diff
@@ -22,9 +22,16 @@ build_opencoarrays()
...
+  fi
   info "Configuring OpenCoarrays in ${PWD} with the command:"
-  info "CC=\"${CC}\" FC=\"${FC}\" $CMAKE \"${opencoarrays_src_dir}\" -DCMAKE_INSTALL_PREFIX=\"${install_path}\" -DMPI_C_COMPILER=\"${MPICC}\" -DMPI_Fortran_COMPILER=\"${MPIFC}\""
-  CC="${CC}" FC="${FC}" $CMAKE "${opencoarrays_src_dir}" -DCMAKE_INSTALL_PREFIX="${install_path}" -DMPI_C_COMPILER="${MPICC}" -DMPI_Fortran_COMPILER="${MPIFC}"
+  info "CC=\"${CC}\" FC=\"${FC}\" $CMAKE \"${opencoarrays_src_dir}\" -DCMAKE_INSTALL_PREFIX=\"${install_path}\" -DMPIEXEC=\"${MPIEXEC}\" -DMPI_C_COMPILER=\"${MPICC}\" -DMPI_Fortran_COMPILER=\"${MPIFC}\""
+  CC="${CC}" FC="${FC}" $CMAKE "${opencoarrays_src_dir}" -DCMAKE_INSTALL_PREFIX="${install_path}" -DMPIEXEC="${MPIEXEC}" -DMPI_C_COMPILER="${MPICC}" -DMPI_Fortran_COMPILER="${MPIFC}"
```
Comment (zbeekman) [inline prerequisites/install-functions/build_opencoarrays.sh]: Tell CMake which `MPIEXEC` we want to use. Passing `MPI_C_COMPILER` should be redundant, but it could be possible that multiple wrappers are in the MPI `bin` dir.

**Comment by zbeekman** [comment]
One caveat is that this needs to be tested on some more exotic systems, like Cray... and with different and more exotic MPI environments. There may be edge cases on that break this...

**Comment by zbeekman** [comment]
@rouson feel free to merge if tests pass and you're happy with this.

**Comment by zbeekman** [comment]
LGTM

Self approving since no one wants to review my code 😢 😝
